graviton/abi_strategy.py

```diff
@@ -183,3 +204,162 @@ def get(self) -> PyTypes:
+            handle adding the method selector so this param.
+            But if set False: when providing `inputs`
+            ensure that the 0'th argument for method calls is the selector.
+            And when set True: when NOT providing `inputs`, the selector arg
+            at index 0 will be added automatically.
+
+        abi_args_mod (optional) - when desiring to mutate the args, provide an ABIArgsMod value
+        """
+        self.contract: abi.Contract = abi.Contract.from_json(contract)
+        self.argument_strategy: Type[ABIStrategy] = argument_strategy
+        self.num_dryruns = num_dryruns
+        self.handle_selector = handle_selector
+        self.abi_args_mod = abi_args_mod
+
+    def abi_method(self, method: Optional[str]) -> abi.Method:
+        assert method, "cannot get abi.Method for bare app call"
+
+        return self.contract.get_method_by_name(method)
+
+    def method_signature(self, method: Optional[str]) -> Optional[str]:
+        """Returns None, for a bare app call (method=None signals this)"""
+        if method is None:
+            return None
+
+        return self.abi_method(method).get_signature()
+
+    def method_selector(self, method: Optional[str]) -> bytes:
+        assert method, "cannot get method_selector for bare app call"
+
+        return self.abi_method(method).get_selector()
+
+    def argument_types(self, method: Optional[str]) -> List[abi.ABIType]:
+        """
+        Argument types (excluding selector)
+        """
+        if method is None:
+            return []
+
+        return [cast(abi.ABIType, arg.type) for arg in self.abi_method(method).args]
+
+    def num_args(self, method: Optional[str]) -> int:
+        return len(self.argument_types(method))
+
+    def generate_inputs(self, method: Optional[str]) -> List[Sequence[PyTypes]]:
+        """
+        Generates inputs appropriate for bare app calls and method calls
+        according to available argument_strategy.
+        """
+        assert (
+            self.argument_strategy
+        ), "cannot generate inputs without an argument_strategy"
+
+        mutating = self.abi_args_mod is not None
+
+        if not (method or mutating):
+            # bare calls receive no arguments (unless mutating)
+            return [tuple() for _ in range(self.num_dryruns)]
+
+        arg_types = self.argument_types(method)
+
+        prefix: List[bytes] = []
+        if self.handle_selector and method:
+            prefix = [self.method_selector(method)]
+
+        modify_selector = False
+        if (action := self.abi_args_mod) in (
+            ABIArgsMod.selector_byte_delete,
+            ABIArgsMod.selector_byte_insert,
+            ABIArgsMod.selector_byte_replace,
+        ):
+            assert (
+                prefix
+            ), f"{self.abi_args_mod=} which means we need to modify the selector, but we don't have one available to modify"
+            modify_selector = True
+
+        def selector_mod(prefix):
```

Comment:
nit: suggested renaming `selector_modification`

Reply:
This is an understandable nit. However, I would have to change the usage of "mod" in several places including in pyteal. So I politely push back but am open to further discussion.

Reply:
sound fine with me, adding a line of comment should be suffice for me, same applies to comment below

Reply:
lgtm